---
title: "Identification of differentially expressed isoforms using Ballgown in _S.namaysuch_ non-parasitized/parasitized liver tissue in two different subspecies: lean and siscowet."
author: "Sam White"
date: "08/17/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Use [Ballgown](https://github.com/alyssafrazee/ballgown) for identification of differentially expressed isoforms in _S.namaysuch_ non-parasitized/parasitized liver tissue in two different subspecies: lean and siscowet.

REQUIRES Linux-based system to run all chunks properly; some chunks will not work on Mac OS!

REQUIRES the following Bash programs:

- `wget`

- `tree`

- `md5sum`

REQUIRES the following R libraries:

- [`Ballgown`](https://github.com/alyssafrazee/ballgown) (Bioconductor)

- `tidyverse`

## Load `R` libraries

```{r}
library("ballgown")
library("tidyverse")
library("ggplot2")
library("Rfast")
```


## Set user variables!
```{r set-pval-qval-cutoffs}
# Set maximum pvalue for isoform expression cutoff
pvalue <- 0.05
# Set maximum qvalue (false discovery rate) for isoform expression cutoff
qvalue <- 0.05
```

## Download Ballgown input files.

Notebooks detailing their creation:

- [FastQ trimming](https://robertslab.github.io/sams-notebook/2022/07/06/SRA-Data-S.namaycush-SRA-BioProject-PRJNA674328-Download-and-QC.html)

- [Genome indexing, and exon/splice sites with HISAT2](https://robertslab.github.io/sams-notebook/2022/08/10/Splice-Site-Identification-S.namaycush-Liver-Parasitized-and-Non-Parasitized-SRA-RNAseq-Using-Hisat2-Stingtie-with-Genome-GCF_016432855.1.html)

- [Mapping and identificaion of isoforms with StingTie](https://robertslab.github.io/sams-notebook/2022/08/10/Splice-Site-Identification-S.namaycush-Liver-Parasitized-and-Non-Parasitized-SRA-RNAseq-Using-Hisat2-Stingtie-with-Genome-GCF_016432855.1.html)

```{bash download-bgown-input-tables}
# Download Ballgown input files and directory structure
wget \
--directory-prefix ../data/ballgown \
--recursive \
--no-check-certificate \
--continue \
--cut-dirs 2 \
--no-host-directories \
--no-parent \
--quiet \
--reject "concatenated-fastq-checksums.md5,original-fastq-checksums.md5" \
--reject-regex "/multiqc_data/" \
--accept "*.ctab,*checksums.md5" https://gannet.fish.washington.edu/Atumefaciens/20220810-snam-hisat2-GCF_016432855.1_index-align-stringtie_isoforms/
```

```{bash cleanup}
rm ../data/ballgown/checksums.md5
```


## Verify checksums

NOTE: Warnings are expected, as the checksums files have checksums for files that are not downloaded for this project.
```{bash}
cd ../data/ballgown

pwd
# Make a line
line="-----------------------------------------------------------------------------------------------"
# Set working directory
wd=$(pwd)
# Loop through directories and verify checksums
for directory in */
do
  cd "${directory}"
  # Get sample name; strips trailing slash from directory name
  sample="${directory%/}"
  
  echo ${line}
  echo "${sample}"
  echo ""
  
  # Confirm checksums; sorts for easier reading
  md5sum --check "${sample}"-checksums.md5 | sort -V
  echo ""
  echo "${line}"
  echo ""
  cd ${wd}
done
# Show downloaded directories/files
tree
```

## Find Ballgown installation location
```{r set-ballgown-install-dir}
data_directory <-  system.file('extdata', package='ballgown') # automatically finds ballgown's installation directory
# examine data_directory:
data_directory
```

## Read in _S.namaycush_ genes BED file
```{r read-in-genes-BED}
genes_BED <- read.table(file = "../data/20220818-snam-GCF_016432855.1_SaNama_1.0_genes.bed")
# Add BED column names for more clarity
colnames(genes_BED) <- c("chr", "start", "end", "name", "score", "strand")
head(genes_BED)
```
## Create Ballgown object
```{r create-ballgown-objects}
# Uses regular expression in samplePattern to find all pertinent folders
# Load all measurement data
bg <- ballgown(dataDir="../data/ballgown/", samplePattern='[NP]*', meas='all')
bg
bg.nonparasitized <- ballgown(dataDir="../data/ballgown/", samplePattern='NP*', meas='all')
bg.nonparasitized
bg.nonparasitized.lean <- ballgown(dataDir="../data/ballgown/", samplePattern='NPLL*', meas='all')
bg.nonparasitized.lean
bg.nonparasitized.siscowet <- ballgown(dataDir="../data/ballgown/", samplePattern='NPSL*', meas='all')
bg.nonparasitized.siscowet

bg.parasitized <- ballgown(dataDir="../data/ballgown/", samplePattern='P*', meas='all')
bg.parasitized
bg.parasitized.lean <- ballgown(dataDir="../data/ballgown/", samplePattern='PLL*', meas='all')
bg.parasitized.lean
bg.parasitized.siscowet <- ballgown(dataDir="../data/ballgown/", samplePattern='PSL*', meas='all')
bg.parasitized.siscowet
```

## Download and filter metadata file

Filtered metadata will be used to create a phenotype dataframe needed for Ballgown differential expression analysis.

Will use the 'strain` column.

`strain` column explanation:

lean parasitized = 1
lean nonparasitized = 2
siscowet parasitized = 3
siscowet nonparasitized = 4

```{r create-dataframes-for-ballgwon-pData}
# Read in metadata file from URL
# Data is alreayd properly sorted (by library) to match directory structure - required by Ballgown.
sample_metadata_full <- read.csv("https://raw.githubusercontent.com/RobertsLab/project-lake-trout/main/data/ballgown-metadata.csv")

# View full metadata
head(sample_metadata_full)

# Create modified metadata for LEAN only
# 1 = LEAN PARASTIZED
# 2 = LEAN NONPARASITIZED
sample_metadata_subset_lean_only <- sample_metadata_full %>% filter(strain == "1" | strain == "2")

# Test to make sure expected number of rows (does NOT count header row)
# Expect 12 rows; 6 samples from each grouping
stopifnot(12 == nrow(sample_metadata_subset_lean_only))
head(sample_metadata_subset_lean_only)


# Create modified metadata for SISCOWET only
# 3 = SISCOWET PARASITIZED
# 4 = SISCOWET NONPARASITIZED
sample_metadata_subset_siscowet_only <- sample_metadata_full %>% filter(strain == "3" | strain == "4")

# Test to make sure expected number of rows (does NOT count header row)
# Expect 12 rows; 6 samples from each grouping
stopifnot(12 == nrow(sample_metadata_subset_siscowet_only))
head(sample_metadata_subset_siscowet_only)

# Create modified metadata for PARASITIZED (leand and siscowet combined)
# 1 = LEAN PARASTIZED
# 3 = SISCOWET PARASITIZED
sample_metadata_subset_parasitized <- sample_metadata_full %>% filter(strain == "1" | strain == "3")

# Test to make sure expected number of rows (does NOT count header row)
# Expect 12 rows; 6 samples from each grouping
stopifnot(12 == nrow(sample_metadata_subset_parasitized))
head(sample_metadata_subset_parasitized)



# Create modified metadata for NONPARASITIZED (leand and siscowet combined)
# 1 = LEAN NONPARASTIZED
# 4 = SISCOWET NONPARASITIZED
sample_metadata_subset_nonparasitized <- sample_metadata_full %>% filter(strain == "3" | strain == "4")

# Test to make sure expected number of rows (does NOT count header row)
# Expect 12 rows; 6 samples from each grouping
stopifnot(12 == nrow(sample_metadata_subset_nonparasitized))
head(sample_metadata_subset_nonparasitized)

```


## Load phenotype dataframe into Ballgown object
```{r load-phenotype-full}
# Load phenotype info into Ballgown
pData(bg) <- sample_metadata_full

# Examine phenotype data as it exists in Ballgown
phenotype_table <-  pData(bg)
head(phenotype_table)
```

## Look at exon info
```{r all-exon-info}
# Exon info
structure(bg)$exon
```

## Look at intron info
```{r all-intron-info}
# Intron info
structure(bg)$intron
```

## Look at transcript (isoform) info
```{r all-transcript-info}
# Transcript info
structure(bg)$trans
```


## Load all transcript expression data
```{r load-all-transcript-expression}
# Expression data
whole_tx_table <-  texpr(bg, 'all')

# Rename gene_names listed as a "."
whole_tx_table <- whole_tx_table %>% mutate(gene_name = ifelse(gene_name == ".", t_name, gene_name))
head(whole_tx_table)

# FPKM values for all transcripts
# Converts output to data frame
transcript_fpkm <- as.data.frame(texpr(bg, 'FPKM'))

# Put rownames into column named "t_id" for further manipulation
transcript_fpkm <- rownames_to_column(transcript_fpkm, "t_id")
head(transcript_fpkm)
```

```{r write-all-fpkm-to-file}
write.csv(transcript_fpkm,
          file = "../data/fpkm-all_transcripts.csv",
          quote = FALSE,
          row.names = FALSE)
```

```{r write-tx-table-to-file}
# Write whole_tx_table to file
write.csv(whole_tx_table,
          file ="../data/whole_tx_table.csv",
          quote = FALSE,
          row.names = FALSE)
```

## Load all gene expression data
```{r load-gene-expression-data}
whole_gx_table <-  gexpr(bg)
head(whole_gx_table)
```